pylint/checkers/variables.py

```diff
@@ -979,267 +992,258 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
             current_consumer = self._to_consume[i]
 
-            # The list of base classes in the class definition is not part
-            # of the class body.
-            # If the current scope is a class scope but it's not the inner
-            # scope, ignore it. This prevents to access this scope instead of
-            # the globals one in function members when there are some common
-            # names.
-            if current_consumer.scope_type == "class" and (
-                utils.is_ancestor_name(current_consumer.node, node)
-                or (i != start_index and self._ignore_class_scope(node))
-            ):
+            # Certain nodes shouldn't be checked as they get checked another time
+            if self._node_should_be_skipped(current_consumer, node, i == start_index):
                 continue
 
-            # Ignore inner class scope for keywords in class definition
-            if (
-                current_consumer.scope_type == "class"
-                and isinstance(node.parent, nodes.Keyword)
-                and isinstance(node.parent.parent, nodes.ClassDef)
-            ):
-                continue
+            action, found_nodes = self._check_consumer(
+                node, stmt, frame, current_consumer, i, base_scope_type
+            )
 
-            # if the name node is used as a function default argument's value or as
-            # a decorator, then start from the parent frame of the function instead
-            # of the function frame - and thus open an inner class scope
-            if (
-                current_consumer.scope_type == "function"
-                and self._defined_in_function_definition(node, current_consumer.node)
-            ):
-                # ignore function scope if is an annotation/default/decorator, as not in the body
+            if action is VariableVisitConsumerAction.CONTINUE:
                 continue
+            if action is VariableVisitConsumerAction.CONSUME:
+                current_consumer.mark_as_consumed(node.name, found_nodes)
+            if action in {
+                VariableVisitConsumerAction.RETURN,
+                VariableVisitConsumerAction.CONSUME,
+            }:
+                return
 
-            if current_consumer.scope_type == "lambda" and utils.is_default_argument(
-                node, current_consumer.node
-            ):
-                continue
+        # we have not found the name, if it isn't a builtin, that's an
+        # undefined name !
+        if self._undefined_variable_is_enabled and not (
+            node.name in nodes.Module.scope_attrs
+            or utils.is_builtin(node.name)
+            or node.name in self.config.additional_builtins
+            or (
+                node.name == "__class__"
+                and isinstance(frame, nodes.FunctionDef)
+                and frame.is_method()
+            )
+            and not utils.node_ignores_exception(node, NameError)
+        ):
```

Comment:
I think I know why : it's due to the ``and`` being before the last parenthesis when it should be after it (adding a parenthesis at the beginning of the if helps).

```suggestion
        if (
            self._undefined_variable_is_enabled and not (
                node.name in nodes.Module.scope_attrs
                or utils.is_builtin(node.name)
                or node.name in self.config.additional_builtins
                or (
                    node.name == "__class__"
                    and isinstance(frame, nodes.FunctionDef)
                    and frame.is_method()
                )
            and not utils.node_ignores_exception(node, NameError)
        ):
```